Modularize iptables manager

pippolo84 opened the pull request:
Modularize the iptables configuration manager in a cell.
Direct dependency to the DaemonConfig global object is avoided through a SharedConfig privately provided to the cell. This config gets the values from all the referenced DaemonConfig helpers at startup, in order to be used by the new cell.

Also, cells depending on iptables (like the node manager one) now takes a reference to it through dependency injection instead of directly calling side effectful exported functions.

~Depends on https://github.com/cilium/cilium/pull/28713~

pippolo84 committed:
iptables: Add a cell for iptables config manager

Modularize iptables configuration manager into its own cell.

The previous Init method is replaced with an Invoke function in order to
be executed when starting the iptables manager cell.

Also, enableIPForwarding is moved in the iptables package and kept
unexported, since it is not used outside of the package.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Rename IptablesManager to Manager

Fix iptables manager name to avoid the repetion of "iptables" in both
package and type names (as suggested in "Effective Go" guidelines about
naming: https://go.dev/doc/effective_go#names)

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/modules/cell.go:
```diff
@@ -0,0 +1,13 @@
+
+var Cell = cell.Module(
+	"modules-manager",
+	"modules manager to search and load kernel modules",
```
No need to repeat the module id in the title, e.g. you could just have "Loads kernel modules required by Cilium".
This package also should be relocated to `pkg/datapath/linux/modules`.

pippolo84 replied:
Fixed in https://github.com/cilium/cilium/pull/28713 as well as for the other comments regarding modules manager.

joamaki commented on pkg/modules/modules.go:
```diff
@@ -50,3 +40,20 @@ func (m *ModulesManager) FindOrLoadModules(expectedNames ...string) error {
+
+	lc.Append(hive.Hook{
+		OnStart: func(_ hive.HookContext) error {
+			modulesList, err := listModules()
```
bit of code golf:
```
func (hive.HookContext) (err error) {
  m.modulesList, err = listModules()
  return
}
```
or maybe even better would be just:

```
func (m *Manager) Start(hive.HookContext) (err error) {
  m.moduleList, err = listModules()
  return
}

func newManager(...) ... {
  ...
  lc.Append(hive.Hook{OnStart: m.Start})
}
```

joamaki commented on pkg/modules/cell.go:
```diff
@@ -0,0 +1,13 @@
+import "github.com/cilium/cilium/pkg/hive/cell"
+
+var Cell = cell.Module(
+	"modules-manager",
```
Maybe `kernel-module-loader` would be more informative?

joamaki commented on pkg/datapath/iptables/cell.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+var Cell = cell.Module(
+	"iptables-manager",
```
I would just call this `iptables` as `-manager` doesn't really add anything.

joamaki commented on pkg/datapath/iptables/cell.go:
```diff
@@ -0,0 +1,46 @@
+
+var Cell = cell.Module(
+	"iptables-manager",
+	"iptables manager to handle iptables-related configuration for Cilium",
```
No need to repeat module id in title

joamaki commented on pkg/datapath/cells.go:
```diff
@@ -57,6 +57,20 @@ var Cell = cell.Module(
 
+	// iptables configuration manager.
+	iptables.Cell,
+	cell.Provide(func(
```
I would move this into `iptables.Cell` and keep things fairly tight here.

joamaki commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -267,46 +275,73 @@ type Manager struct {
+// Start initializes the iptables manager and checks for iptables kernel modules availability.
+func (m *Manager) Start(ctx hive.HookContext) error {
+	if err := enableIPForwarding(); err != nil {
+		log.Fatalf("enabling IP forwarding via sysctl failed: %s", err)
```
Don't fatal, but return an error instead.

joamaki commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -267,46 +275,73 @@ type Manager struct {
+		"ip6_tables", "ip6table_mangle", "ip6table_raw", "ip6table_filter",
+	); err != nil {
+		if m.sharedCfg.EnableIPv6 {
+			return errors.New(
```
The error is now dropped, I would wrap it instead with fmt.Errorf and `%w`.

pippolo84 committed:
iptables: Avoid depending from DaemonConfig directly

iptables has been modularized into a cell, but still depends directly on
option.DaemonConfig global variables and helpers.

The configuration flags that are "private" to the cell are provided
through cell.Config.
Regarding the flags and the helpers shared with other parts of the code,
a SharedConfig struct is provided in the cell itself.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Expose ipset funcs as Manager methods

Move AddToNodeIpset and RemoveFromNodeIpset into the iptables manager,
so that other cells depending on them will import the iptables config
manager through dependency injection.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

dylandreimerink commented on pkg/datapath/iptables/cell.go:
```diff
@@ -0,0 +1,95 @@
+	"Handle iptables-related configuration for Cilium",
+
+	cell.Config(defaultConfig),
+	cell.Provide(func(
```
This is essentially just a layer around the daemonConfig, I don't see the need to expose this as a separate type in hive. I understand it makes sense internally for the purposes of mocking though

pippolo84 replied:
Yep, main reason to do that is to avoid depending on the big global config object `DaemonConfig`. But it's better to provide it only privately. I'll fix this, thanks :+1:

mhofstetter commented on pkg/datapath/iptables/cell.go:
```diff
@@ -0,0 +1,94 @@
+
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Duration(IPTablesLockTimeout, def.IPTablesLockTimeout, "Time to pass to each iptables invocation to wait for xtables lock acquisition")
+	flags.StringSlice(DisableIptablesFeederRules, def.DisableIptablesFeederRules, "Chains to ignore when installing feeder rules.")
```
changing this to a Hive config no longer parses the passed string as space-separated string. Instead we have to adapt the Helm Chart to pass a comma-separated string -> https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-configmap.yaml#L926

pippolo84 replied:
I've put this on hold to wait for the final decision about how to approach the StringSlice issue: https://github.com/cilium/cilium/pull/28919#issuecomment-1790405427

pippolo84 replied:
I'll take care of this in a follow-up PR as soon as we reach consensus.

mhofstetter commented on pkg/datapath/iptables/cell.go:
```diff
@@ -0,0 +1,94 @@
+	"github.com/cilium/cilium/pkg/time"
+)
+
+const (
```
i'd prefer having these const values inlined. It's still possible to keep the comments above the corresponding flag definitions in `Flags(...)`

joamaki replied:
Yep likewise. I would move the doc comments on to the struct fields.

dylandreimerink commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -267,46 +274,74 @@ type Manager struct {
+// Start initializes the iptables manager and checks for iptables kernel modules availability.
+func (m *Manager) Start(ctx hive.HookContext) error {
+	if err := enableIPForwarding(); err != nil {
+		return fmt.Errorf("enabling IP forwarding via sysctl failed: %w", err)
```
If we fail to set the sysctl values, it does not necessity mean that Cilium will not function (the correct values may be set but we don't have write access), so I would convert this to a warning like the check below.

dylandreimerink reviewed (approved):
LGTM. Just one nit, but given its copied code will not block on that

joamaki commented on pkg/datapath/cells.go:
```diff
@@ -57,6 +57,9 @@ var Cell = cell.Module(
 	// The modules manager to search and load kernel modules.
 	modules.Cell,
 
+	// iptables configuration manager.
```
Maybe better as something like "Manages Cilium-specific iptables rules"?

mhofstetter reviewed (approved):
Thanks - LGTM!
